src/main/java/com/redhat/devtools/intellij/qute/psi/internal/java/AbstractQuteTemplateLinkCollector.java

```diff
@@ -21,244 +21,262 @@
+                    templateFile, templatePathInfo);
+        } catch (IndexNotReadyException | ProcessCanceledException | CancellationException e) {
+            throw e;
+        } catch (RuntimeException e) {
+            LOGGER.log(Level.WARNING, "Error while creating Qute CodeLens for Java file.", e);
+        }
+    }
+
+    /**
+     * Returns the relative template file path from the module.
+     *
+     * @param templateFilePath the template file path (without file extension).
+     * @param templateFile     the template file and null otherwise.
+     * @param module           the module.
+     * @return the relative template file path from the module.
+     */
+    private @NotNull String getRelativePath(@NotNull String templateFilePath, boolean definedSuffix, @Nullable VirtualFile templateFile, @NotNull Module module) {
+        if (templateFile != null) {
+            // The template file exists
+            // Compute the relative path from the module root path:
+            // - src/main/resources/templates/page.html for Maven project
+            // - resources/templates/page.html for Gradle project
+            VirtualFile @NotNull [] roots = ModuleRootManager.getInstance(module).getContentRoots();
+            Arrays.sort(roots, Comparator.comparingInt(r -> r.getPath().length())); //put root with smallest path first (eliminates generated sources roots)
+            for (VirtualFile root : roots) {
+                String path = VfsUtilCore.getRelativePath(templateFile, root);
+                if (path != null) {
+                    return path;
+                }
+            }
+            // Corner usecase, returns the full template file path.
+            return templateFile.getPath();
+        }
+        // The template file doesn't exist
+        // Compute the relative path from the module root path:
+        // - src/main/resources/templates/page.html for Maven project
+        // - resources/templates/page.html for Gradle project
+        return relativeTemplatesBaseDir + templateFilePath + (definedSuffix ? "" : PREFERRED_SUFFIX);
+    }
+
+
+    private @Nullable VirtualFile getTemplateFile(String templateUri, boolean discoverSuffix) {
+        if (templatesDir == null || !templatesDir.exists()) {
+            return null;
+        }
+        if (!discoverSuffix) {
+            return templatesDir.findFileByRelativePath(templateUri);
+        }
+
+        for (String suffix : suffixes) {
+            VirtualFile templateFile = templatesDir.findFileByRelativePath(templateUri + suffix);
+            if (templateFile != null && templateFile.exists()) {
+                return templateFile;
+            }
+        }
+        return templatesDir.findFileByRelativePath(templateUri + PREFERRED_SUFFIX);
+    }
+
+    protected static String getVirtualFileUrl(Module project, String templateFilePath) {
+        try {
+            for (VirtualFile root : ModuleRootManager.getInstance(project).getContentRoots()) {
```

Comment:
this is where the template URI picks up the generated sources folder, when a template is missing. Needs to be sorted

Reply:
fixed